vscode-trace-extension/src/utils/trace-server-manager.ts

```diff
@@ -0,0 +1,86 @@
+/***************************************************************************************
+ * Copyright (c) 2023 BlackBerry Limited and others.
+ *
+ * Licensed under the MIT license. See LICENSE file in the project root for details.
+ ***************************************************************************************/
+type Handlers = [() => Promise<void>, (() => Promise<void>)?];
+
+export class TraceServerManager {
+
+    /**
+     * Mapping file extensions to handlers
+     */
+    private fileExtsHandlers: Map<string[], Handlers> = new Map;
+    private isManagerDisposed = false;
+    private stopHandlerToUse: (() => Promise<void>) | undefined = undefined;
+
+    /**
+     * Add a new handler (function that runs on adopting extension to start its trace server) to the list
+     *
+     * @param fileExts list of file extensions an extension will be prioritized for
+     * @param startHandler the start handler to register
+     * @param stopHandler the stop handler to register
+     */
+    addServerHandlers(fileExts: string[], startHandler: () => Promise<void>, stopHandler?: () => Promise<void>): void {
+        if (!this.isDisposed()) {
+            this.fileExtsHandlers.set(fileExts, [startHandler, stopHandler]);
+        }
+    }
+
+    /**
+     * Look for appropriate start handler based on the file extension and execute it
+     *
+     * @param fileExt file extension from the trace file url
+     */
+    async startServer(fileExt: string): Promise<void> {
+        if (!this.isDisposed()) {
+            // If fileExt is empty, execute start handler of first adapter extension
+            if (!fileExt) {
+                const firstExtHandlers = this.fileExtsHandlers.entries().next().value;
+                await firstExtHandlers[0]();
+                this.stopHandlerToUse = firstExtHandlers[1];
+                return;
+            }
+            // Execute start handler of first adapter extention that support the file extension
+            const handler = [...this.fileExtsHandlers]
+                .find(([k]) =>
+                    k.includes(fileExt));
+            if (handler) {
+                await handler[1][0]();
+                this.stopHandlerToUse = handler[1][1];
+                return;
+            }
+            // If no extension supports the file extension
+            // execute handler with a wildcard extension
+            const wildcardHandler = [...this.fileExtsHandlers]
+                .find(([k]) =>
+                    k.includes('*'));
+            if (wildcardHandler) {
+                await wildcardHandler[1][0]();
+                this.stopHandlerToUse = wildcardHandler[1][1];
+                return;
+            }
+            // execute handler for extension with no file extension listed if found
+            const noExtHandlers = this.fileExtsHandlers.get([]);
+            noExtHandlers?.[0];
+            this.stopHandlerToUse = noExtHandlers?.[1];
+        }
+    }
```

Comment:
- I'd have a concern about how to formally test this method above.
- And on a similar note, should we consider some PR to make the [vscode-trace-server](https://github.com/eclipse-cdt-cloud/vscode-trace-server) extension register itself as handlers by default (if living beside this very extension)?
- These questions would also be for @bhufmann; thanks.

Reply:
> * I'd have a concern about how to formally test this method above.
* We might be able to test the API by mocking an adopting extension and mock opening trace files

> * And on a similar note, should we consider some PR to make the [vscode-trace-server](https://github.com/eclipse-cdt-cloud/vscode-trace-server) extension register itself as handlers by default (if living beside this very extension)?
*  [vscode-trace-server](https://github.com/eclipse-cdt-cloud/vscode-trace-server) could register itself as a wildcard adopting extension. Do you want it to be prioritized over all other adopting extensions?